File: examples/multi_branch_pipeline/python/multi_branch_pipeline.py
"""
SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""  # noqa: E501

import time
from argparse import ArgumentParser

from holoscan.conditions import CountCondition, PeriodicCondition
from holoscan.core import Application, ConditionType, IOSpec, Operator, OperatorSpec
from holoscan.schedulers import EventBasedScheduler, GreedyScheduler, MultiThreadScheduler


class PingTxOp(Operator):
    """Simple transmitter operator.

    This operator has:
        outputs: "out"

    On each tick, it transmits an integer on the "out" port. The value on the first call to
    compute is equal to `initial_value` and it increments by `increment` on each subsequent
    call.
    """

    def __init__(self, fragment, *args, initial_value=0, increment=0, **kwargs):
        self.count = 0
        self.initial_value = initial_value
        self.increment = increment

        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        # Note: Setting ConditionType.NONE overrides the default of
        #   ConditionType.DOWNSTREAM_MESSAGE_AFFORDABLE. This means that the operator will be
        #   triggered regardless of whether any operators connected downstream have space in their
        #   queues.
        spec.output("out").condition(ConditionType.NONE)

    def compute(self, op_input, op_output, context):
        value = self.initial_value + self.count * self.increment
        op_output.emit(value, "out")
        self.count += 1


class IncrementOp(Operator):
    """Add a fixed value to the input and transmit the result."""

    def __init__(self, fragment, *args, increment=0, **kwargs):
        self.increment = increment

        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        """Setup the input and output ports with custom settings on the input port.

        Notes
        -----
        For policy:
          - IOSpec.QueuePolicy.POP = pop the oldest value in favor of the new one when the queue
            is full
          - IOSpec.QueuePolicy.REJECT = reject the new value when the queue is full
          - IOSpec.QueuePolicy.FAULT = fault if queue is full (default)

        For capacity:
          When capacity > 1, even once messages stop arriving, this entity will continue to
          call ``compute`` for each remaining item in the queue.

        The ``condition`` method call here is the same as the default setting, and is shown
        only for completeness. `min_size` = 1 means that this operator will not call compute
        unless there is at least one message in the queue.

        One could also set the receiver's capacity and policy via the connector method:

            .connector(
                IOSpec.ConnectorType.DOUBLE_BUFFER,
                capacity=1,
                policy=1,  # 1 = reject
            )

        but that is less flexible as `IOSpec::ConnectorType::kDoubleBuffer` is appropriate for
        within-fragment connections, but will not work if the operator was connected to a
        different fragment. By passing the capacity and policy as arguments to the
        `OperatorSpec::input` method, the SDK can still select the appropriate default receiver
        type depending on whether the connection is within-fragment or across fragments (or
        whether an annotated variants of the receiver class is needed for data flow tracking).

        Note that if policy was not specific in the `input` call as done here, it would also be
        possible for an application author to set the `policy` after the operator is constructed
        by using the `Operator.queue_policy` property (see commented code in
        `MultiRateApp.compose` below).
        """
        spec.input("in", policy=IOSpec.QueuePolicy.REJECT).condition(
            ConditionType.MESSAGE_AVAILABLE, min_size=1, front_stage_max_size=1
        )

        spec.output("out")

    def compute(self, op_input, op_output, context):
        value = op_input.receive("in")
        new_value = value + self.increment
        op_output.emit(new_value, "out")


class PingRxOp(Operator):
    """Simple (multi)-receiver operator.

    This is an example of a native operator with one input port.
    On each tick, it receives an integer from the "in" port.
    """

    def __init__(self, fragment, *args, **kwargs):
        # Need to call the base class constructor last
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.input("in")

    def compute(self, op_input, op_output, context):
        # In this case, nothing will be printed until all messages have
        # been received.
        value = op_input.receive("in")
        print(f"receiver '{self.name}' received value: {value}")


# Now define a simple application using the operators defined above
class MultiRateApp(Application):
    """This application has a single transmitter connected to two parallel branches

    The geometry of the application is as shown below:

       increment1--rx1
      /
    tx
      \
       increment2--rx2

    The top branch is forced via a PeriodicCondition to run at a slower rate than
    the source. It is currently configured to discard any extra messages that arrive
    at increment1 before it is ready to execute again, but different behavior could be
    achieved via other settings to policy and/or queue sizes.

    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def compose(self):
        # Configure the operators. Here we use CountCondition to terminate
        # execution after a specific number of messages have been sent and a
        # PeriodicCondition to control how often messages are sent.
        source_rate_hz = 60  # messages sent per second
        period_source_ns = int(1e9 / source_rate_hz)  # period in nanoseconds
        tx = PingTxOp(
            self,
            CountCondition(self, 100),
            PeriodicCondition(self, recess_period=period_source_ns),
            initial_value=0,
            increment=1,
            name="tx",
        )

        # first branch will have a periodic condition so it can't run faster than 5 Hz
        branch1_hz = 5
        period_ns1 = int(1e9 / branch1_hz)
        increment1 = IncrementOp(
            self,
            PeriodicCondition(self, recess_period=period_ns1),
            name="increment1",
        )
        # could set the queue policy here if it wasn't already set to REJECT in IncrementOp.start
        # increment1.queue_policy(
        #     "in", port_type=IOSpec.IOType.INPUT, policy=IOSpec.QueuePolicy.REJECT
        # )

        rx1 = PingRxOp(self, name="rx1")
        self.add_flow(tx, increment1)
        self.add_flow(increment1, rx1)

        # second branch does not have a periodic condition so will tick on every message sent by tx
        increment2 = IncrementOp(
            self,
            name="increment2",
        )
        # could set the queue policy here if it wasn't already set to REJECT in IncrementOp.start
        # increment2.queue_policy(
        #     "in", port_type=IOSpec.IOType.INPUT, policy=IOSpec.QueuePolicy.REJECT
        # )

        rx2 = PingRxOp(self, name="rx2")
        self.add_flow(tx, increment2)
        self.add_flow(increment2, rx2)


def main(threads, event_based):
    app = MultiRateApp()
    if threads == 0:
        # Explicitly setting GreedyScheduler is not strictly required as it is the default.
        scheduler = GreedyScheduler(app, name="greedy_scheduler")
    else:
        scheduler_class = EventBasedScheduler if event_based else MultiThreadScheduler
        scheduler = scheduler_class(
            app,
            worker_thread_number=threads,
            stop_on_deadlock=True,
            stop_on_deadlock_timeout=500,
            name="multithread_scheduler",
        )
    app.scheduler(scheduler)
    tstart = time.time()
    app.run()
    duration = time.time() - tstart
    print(f"Total app runtime = {duration:0.3f} s")


if __name__ == "__main__":
    # Parse args
    parser = ArgumentParser(description="Multi-rate pipeline example")
    parser.add_argument(
        "-t",
        "--threads",
        type=int,
        default=5,
        help=(
            "The number of threads to use for multi-threaded schedulers. Set this to 0 to use "
            "the default greedy scheduler instead. To use the event-based scheduler instead of "
            "the default multi-thread scheduler, please specify --event_based."
        ),
    )
    parser.add_argument(
        "--event_based",
        action="store_true",
        help=(
            "Sets the application to use the event-based scheduler instead of the default "
            "multi-thread scheduler when threads > 0."
        ),
    )

    args = parser.parse_args()
    if args.threads < 0:
        raise ValueError("threads must be non-negative")

    main(threads=args.threads, event_based=args.event_based)

File: examples/multi_branch_pipeline/cpp/multi_branch_pipeline.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <fmt/core.h>
#include <unistd.h>

#include <memory>
#include <string>
#include <vector>

#include "holoscan/holoscan.hpp"

namespace holoscan::ops {

class PingTxOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PingTxOp)

  PingTxOp() = default;

  void setup(OperatorSpec& spec) override {
    // Note: Setting ConditionType::kNone overrides the default of
    //   ConditionType::kDownstreamMessageAffordable. This means that the operator will be triggered
    //   regardless of whether any operators connected downstream have space in their queues.
    spec.output<int64_t>("out").condition(ConditionType::kNone);
    spec.param(initial_value_,
               "initial_value",
               "Initial value",
               "Initial value to emit",
               static_cast<int64_t>(0));
    spec.param(increment_,
               "increment",
               "Increment",
               "Integer amount to increment the value by on each subsequent call",
               static_cast<int64_t>(1));
  }

  void compute([[maybe_unused]] InputContext& op_input, OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    auto value = initial_value_.get() + count_ * increment_.get();
    op_output.emit(value, "out");
    count_ += 1;
  };

 private:
  Parameter<int64_t> initial_value_;
  Parameter<int64_t> increment_;
  int64_t count_ = 0;
};

class IncrementOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(IncrementOp)

  IncrementOp() = default;

  /* Setup the input and output ports with custom settings on the input port.
   *
   *  Notes
   *  =====
   *  For policy:
   *    - IOSpec::QueuePolicy::kPop = pop the oldest value in favor of the new one when the
   *      queue is full
   *    - IOSpec::QueuePolicy::kReject = reject the new value when the queue is full
   *    - IOSpec::QueuePolicy::kFault = fault if queue is full (default)
   *
   *  For capacity:
   *    When capacity > 1, even once messages stop arriving, this entity will continue to
   *    call ``compute`` for each remaining item in the queue.
   *
   *  The ``condition`` method call here is the same as the default setting, and is shown
   *  only for completeness. `min_size` = 1 means that this operator will not call compute
   *  unless there is at least one message in the queue.
   *
   *  One could also set the receiver's capacity and policy via the connector method:
   *
   *          .connector(IOSpec::ConnectorType::kDoubleBuffer,
   *                     Arg("capacity", static_cast<uint64_t>(1)),
   *                     Arg("policy", static_cast<uint64_t>(1)))  // 1 = reject
   *
   *  but that is less flexible as `IOSpec::ConnectorType::kDoubleBuffer` is appropriate for
   *  within-fragment connections, but will not work if the operator was connected to a different
   *  fragment. By passing the capacity and policy as arguments to the OperatorSpec::input method,
   *  the SDK can still select the appropriate default receiver type depending on whether the
   *  connection is within-fragment or across fragments (or whether an annotated variants of the
   *  receiver class is needed for data flow tracking).
   */
  void setup(OperatorSpec& spec) override {
    spec.input<int64_t>("in", IOSpec::IOSize(1), IOSpec::QueuePolicy::kReject)
        .condition(  // arguments to condition here are the same as the defaults
            ConditionType::kMessageAvailable,
            Arg("min_size", static_cast<uint64_t>(1)),
            Arg("front_stage_max_size", static_cast<size_t>(1)));
    spec.output<int64_t>("out");

    spec.param(increment_,
               "increment",
               "Increment",
               "Integer amount to increment the input value by",
               static_cast<int64_t>(0));
  }

  void compute(InputContext& op_input, OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    int64_t value = op_input.receive<int64_t>("in").value();
    // increment value by the specified increment
    int64_t new_value = value + increment_.get();
    op_output.emit(new_value, "out");
  };

 private:
  Parameter<int64_t> increment_;
};

class PingRxOp : public Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PingRxOp)

  PingRxOp() = default;

  void setup(OperatorSpec& spec) override { spec.input<int64_t>("in"); }

  void compute(InputContext& op_input, [[maybe_unused]] OutputContext& op_output,
               [[maybe_unused]] ExecutionContext& context) override {
    auto value = op_input.receive<int64_t>("in").value();
    HOLOSCAN_LOG_INFO("receiver '{}' received value: {}", name(), value);
  };
};

}  // namespace holoscan::ops

/* This application has a single transmitter connected to two parallel branches
 *
 *  The geometry of the application is as shown below:
 *
 *     increment1--rx1
 *    /
 *  tx
 *    \
 *     increment2--rx2
 *
 *  The top branch is forced via a PeriodicCondition to run at a slower rate than
 *  the source. It is currently configured to discard any extra messages that arrive
 *  at increment1 before it is ready to execute again, but different behavior could be
 *  achieved via other settings to policy and/or queue sizes.
 */
class MultiRateApp : public holoscan::Application {
 public:
  void compose() override {
    using namespace holoscan;

    // Configure the operators. Here we use CountCondition to terminate
    // execution after a specific number of messages have been sent and a
    // PeriodicCondition to control how often messages are sent.
    int64_t source_rate_hz = 60;                                // messages sent per second
    int64_t period_source_ns = 1'000'000'000 / source_rate_hz;  // period in nanoseconds
    auto tx = make_operator<ops::PingTxOp>(
        "tx",
        make_condition<CountCondition>("count", 100),
        make_condition<PeriodicCondition>("tx-period", period_source_ns));

    // first branch will have a periodic condition so it can't run faster than 5 Hz
    int64_t branch1_hz = 5;
    int64_t period_ns1 = 1'000'000'000 / branch1_hz;
    auto increment1 = make_operator<ops::IncrementOp>(
        "increment1", make_condition<PeriodicCondition>("increment1-period", period_ns1));
    // could set the queue policy here if it wasn't already set to `IOSpec::QueuePolicy::kReject`
    // in IncrementOp::setup() method as follows:
    //   increment1->queue_policy("in", IOSpec::IOType::kInput, IOSpec::QueuePolicy::kReject);
    auto rx1 = make_operator<ops::PingRxOp>("rx1");
    add_flow(tx, increment1);
    add_flow(increment1, rx1);

    // second branch is the same, but no periodic condition so will tick on every received message
    auto increment2 = make_operator<ops::IncrementOp>("increment2");
    // could set the queue policy here if it wasn't already set to `IOSpec::QueuePolicy::kReject`
    // in IncrementOp::setup() method as follows:
    //   increment2->queue_policy("in", IOSpec::IOType::kInput, IOSpec::QueuePolicy::kReject);
    auto rx2 = make_operator<ops::PingRxOp>("rx2");
    add_flow(tx, increment2);
    add_flow(increment2, rx2);
  }
};

int main([[maybe_unused]] int argc, char** argv) {
  auto app = holoscan::make_application<MultiRateApp>();

  // Get the configuration
  auto config_path = std::filesystem::canonical(argv[0]).parent_path();
  config_path += "/multi_branch_pipeline.yaml";
  app->config(config_path);

  auto scheduler = app->from_config("scheduler").as<std::string>();
  if (scheduler == "multi_thread") {
    // use MultiThreadScheduler instead of the default GreedyScheduler
    app->scheduler(app->make_scheduler<holoscan::MultiThreadScheduler>(
        "multithread-scheduler", app->from_config("multi_thread_scheduler")));
  } else if (scheduler == "event_based") {
    // use EventBasedScheduler instead of the default GreedyScheduler
    app->scheduler(app->make_scheduler<holoscan::EventBasedScheduler>(
        "event-based-scheduler", app->from_config("event_based_scheduler")));
  } else if (scheduler == "greedy") {
    app->scheduler(app->make_scheduler<holoscan::GreedyScheduler>(
        "greedy-scheduler", app->from_config("greedy_scheduler")));
  } else if (scheduler != "default") {
    throw std::runtime_error(fmt::format(
        "unrecognized scheduler option '{}', should be one of {{'multi_thread', 'event_based', "
        "'greedy', 'default'}}",
        scheduler));
  }

  app->run();

  return 0;
}
